### ISSUE TITLE
API Signature route unprotected?

### ISSUE THREAD

**Issue by osseonews** [cond0r/verify-webhooks-with-middleware#1]
Thanks very much for this repo as I was building a similar thing. My only question, because I am facing a similar issue, is does this really protect API routes? B/c to get the signature you need to hit the pages/api/signature. This returns the signature, and so obviously this endpoint itself can't be protected. So what prevents somebody from just hitting the signature endpoint, getting a signature and then attaching it to a request to the webhook api? I mean it seems like you can never really protect API's as there is always 1 api route that must be open to generate the signature,  it would be trivial to just get around everything by hitting this API route to get the signature.  So even if you don't have access to the Secret itself, who cares. You can just generate the signature and access all the protected API routes. how do you solve this basic problem?

**Comment by cond0r** [comment]
@osseonews That signature API page was just a proof of concept to get the signature by hashing that body to simulate a webhook trigger (I was showing this to someone). You usually would hash the body before triggering the webhook, and  with middleware, you verify the signature.

**Comment by osseonews** [comment]
Right, I got that, but let's say we were sending a request to a protected API route from the client (not as a server to server webhook), for example within our NextJs App to an API route. So then we need a way to add to a signature to the API request in the client. So only way to do that, without exposing the secret in the client, is to first send a request to another backend API to generate the signature and then attach the signature in the 2nd request. But, that means that the 1st request endpoint is unprotected even if the client doesn't expose the secret. So then what is the point of actually securing the API route, as anyone can get the signature by hitting the 1st endpoint?

Isn't that correct? So there is no way of securing endpoints from the client side in the NextJS app? Am I missing something here? Protected API's only works for an external API request server to server?

**Comment by cond0r** [comment]
Why don't you send it to the unprotected endpoint, hash it there and forward it to the protected one, so you never return the signature to the client? And for the unprotected routes, if is just client -> API routes, enable [cors](https://vercel.com/guides/how-to-enable-cors) to be sure requests come only from your client app.

**Comment by osseonews** [comment]
Thanks.  questions below.
"hash it there and forward it to the protected one" - Hmm, I wasn't aware you can forward. We are using Edge functions. Can I just call a 2nd Edge function from within another one?

"CORS" - I kind of view CORS as useless. B/c if some malicious actor knows your API route, they will just call it via a server or command line, or whatever, and bypass CORS. They won't call it from the client. I honestly don't know what the point of CORS is. It doesn't protect anything really.

**Comment by cond0r** [comment]
Yes, you make a fetch request to the other endpoint, but you are again in the same concern with the command line to bypass.
